### PR TITLE
Fix unsafe publication of invalid license enforcer

### PR DESCRIPTION
**Pull request by jasontedor**
The invalid license enforced is exposed to the cluster state update thread (via the license state listener) before the constructor has finished. This violates the JLS for safe publication of an object, and means there is a concurrency bug lurking here. This commit addresses this by avoiding publication of the invalid license enforcer before the constructor has returned.

Relates #40957

### CODE REVIEW

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java [hunk 1]
```diff
@@ -461,8 +461,10 @@ protected Clock getClock() {
         MlLifeCycleService mlLifeCycleService = new MlLifeCycleService(environment, clusterService, datafeedManager,
                 autodetectProcessManager, memoryTracker);
 
-        // This object's constructor attaches to the license state, so there's no need to retain another reference to it
-        new InvalidLicenseEnforcer(getLicenseState(), threadPool, datafeedManager, autodetectProcessManager);
+        // this object's attaches to the license state, and never needs to be removed, so there's no need to retain another reference to it
```
Comment (jpountz) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java]: should it be `s/object's/object/`?

**Review by jpountz — APPROVED** [review]
LGTM but please wait for @droberts195 to have a chance to have a look before merging.

**Review by droberts195 — APPROVED** [review]
LGTM

Thanks for fixing this

**Pull request merged**